spring-session/src/main/java/org/springframework/session/data/mongo/MongoSessionRepository.java

```diff
@@ -0,0 +1,160 @@
+		session.setMaxInactiveIntervalInSeconds(maxInactiveIntervalInSeconds);
+		return session;
+	}
+
+	@Override
+	public void save(ExpiringSession session) {
+		DBObject jo = sessionSerializer.serializeSession(session);
+		mongoOperations.getCollection(collectionName).save(jo);
+		eventPublisher.publishEvent(new SessionCreatedEvent(this, session.getId()));
+	}
+
+	@Override
+	public ExpiringSession getSession(String id) {
+		DBObject sessionWrapper = findSession(id);
+		if (sessionWrapper == null) {
+			return null;
+		}
+		ExpiringSession session = sessionSerializer.deserializeSession(sessionWrapper);
+		if (session.isExpired()) {
+			expireSession(id);
+			return null;
+		}
+		return session;
+	}
+
+	/**
+	 * Currently this repository allows only querying against {@code PRINCIPAL_NAME_INDEX_NAME}
+	 *
+	 * @param indexName  the name if the index (i.e. {@link FindByIndexNameSessionRepository#PRINCIPAL_NAME_INDEX_NAME})
+	 * @param indexValue the value of the index to search for.
+	 * @return sessions map
+	 */
+	@Override
+	public Map<String, ExpiringSession> findByIndexNameAndIndexValue(String indexName, String indexValue) {
+		if (!PRINCIPAL_NAME_INDEX_NAME.equals(indexName)) {
+			return Collections.emptyMap();
+		}
+		HashMap<String, ExpiringSession> result = new HashMap<String, ExpiringSession>();
+		List<DBObject> mapSessions = mongoOperations.find(getPrincipalQuery(indexValue), DBObject.class, collectionName);
+		for (DBObject session : mapSessions) {
+			ExpiringSession mapSession = sessionSerializer.deserializeSession(session);
+			result.put(mapSession.getId(), mapSession);
+		}
+		return result;
+	}
+
+	@Override
+	public void delete(String id) {
+		mongoOperations.remove(findSession(id), collectionName);
+		eventPublisher.publishEvent(new SessionDeletedEvent(this, id));
+	}
+
+	/**
+	 * Method cleanups expired sessions and emits {@code SessionExpiredEvent} event for expiring sessions
+	 */
+	@Scheduled(cron = "0 * * * * *")
+	public void cleanupExpiredSessions() {
+		long now = System.currentTimeMillis();
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Cleaning up sessions expiring at ", new Date(now));
+		}
+
+		long threshold = now - TimeUnit.SECONDS.toMillis(maxInactiveIntervalInSeconds);
+
+		List<DBObject> expiredSessions = mongoOperations.find(
+				Query.query(Criteria.where(sessionSerializer.getLastAccessedFieldName()).lt(threshold)), DBObject.class, collectionName);
+
+		for (DBObject expiredSession : expiredSessions) {
+			expireSession((String) expiredSession.get(ID_FIELD_NAME));
+		}
+	}
+
+	private Query getPrincipalQuery(String indexValue) {
+		return Query.query(Criteria.where(sessionSerializer.getPrincipalFieldName()).is(indexValue));
+	}
+
+	DBObject findSession(String id) {
+		return mongoOperations.findById(id, DBObject.class, collectionName);
+	}
+
+	private void expireSession(String id) {
+		mongoOperations.remove(findSession(id), collectionName);
+		eventPublisher.publishEvent(new SessionExpiredEvent(this, id));
```

Comment:
Unfortunately this event will not be published across all nodes. This means that we don't really get the expected support. Perhaps (at least for now) Mongo support just doesn't support events (this is allowed).